File: IssueTracker.Data/Repositories/IssueRepository.cs
using System.Collections.Generic;
using System.Linq;
using IssueTracker.Common.Data.Repositories;
using IssueTracker.Common.Models;
using IssueTracker.Common.ViewModels;

namespace IssueTracker.Data.Repositories
{
	public class IssueRepository : BaseProjectRepository<Issue>, IIssueRepository
	{
		public IEnumerable<Issue> Search(Search search, Sort sort)
		{
			var issues = (IEnumerable<Issue>) Context.Issues.Include("Developer").Include("Tester").Include("Priority").Include("Status").Include("Project");
			ApplyFilter(ref issues, search);
			ApplySort(ref issues, sort);
			return issues.Where(x => !x.IsDeleted).Skip(search.start - 1).Take(search.end - search.start + 1);
		}

		public int HighestNumber(Project project)
		{
			var issue = Context.Issues.Where(x => x.Project.Id == project.Id).OrderByDescending(x => x.Number).FirstOrDefault();
			return issue == null ? 0 : issue.Number;
		}

		private void ApplySort(ref IEnumerable<Issue> issues, Sort sort)
		{
			if (sort.direction == SortDirection.Ascending)
				issues = issues.OrderBy(sort.GetComparerFunction()).ThenBy(x => x.Opened);
			else if (sort.direction == SortDirection.Descending)
				issues = issues.OrderByDescending(sort.GetComparerFunction()).ThenBy(x => x.Opened);
		}

		private void ApplyFilter(ref IEnumerable<Issue> issues, Search search)
		{
			issues = issues.Where(x => x.Project.Id == search.project.Id);
			if (search.milestones.Any())
				issues = issues.Where(x => search.milestones.Select(y => y.Id).Contains(x.Milestone.Id));
			if (search.priorities.Any())
				issues = issues.Where(x => search.priorities.Select(y => y.Id).Contains(x.Priority.Id));
			if (search.statuses.Any())
				issues = issues.Where(x => search.statuses.Select(y => y.Id).Contains(x.Status.Id));
			if (search.developers.Any())
				issues = issues.Where(x => search.developers.Select(y => y.Id).Contains(x.Developer.Id));
			if (search.testers.Any())
				issues = issues.Where(x => search.testers.Select(y => y.Id).Contains(x.Tester.Id));
			if (search.status != null)
				issues = issues.Where(x => x.Status.Id == search.status.Id);
			if (search.developer != null)
				issues = issues.Where(x => x.Developer.Id == search.developer.Id);
			if (search.tester != null)
				issues = issues.Where(x => x.Tester.Id == search.tester.Id);
			if (!string.IsNullOrEmpty(search.filter))
			{
				search.filter = search.filter.Trim().ToLower();
				issues = issues.Where(x =>
					x.Name.ToLower().Trim().Contains(search.filter) || x.Comments.Trim().ToLower().Contains(search.filter) ||
					x.Status.ToString().Trim().ToLower().Contains(search.filter) || x.Priority.ToString().Contains(search.filter) ||
					x.Developer.ToString().Trim().ToLower().Contains(search.filter) ||
					x.Tester.ToString().Trim().ToLower().Contains(search.filter));
			}
		}
	}
}

File: data/repositories/issueRepository.js
var Promise = require("bluebird");
var config = require("../../config");
var repositories = require("../repositories");

var repository = Object.spawn(require("./baseRepository"), {
	model: require("../models").Issue
});

repository.search = function(projectId, filter, sortDirection, sortComparer, start, end) {
	console.time("search");
	return repository.get(_buildParameters(projectId, filter), {
		sort: _buildSort(sortDirection, sortComparer),
		skip: start - 1,
		limit: end - start + 1,
		projection: {
			name: true,
			details: true,
			number: true,
			priorityId: true,
			developerId: true
		}
	}).then(function(issues) {
		console.timeEnd("search");
		return issues;
	});

	function _buildParameters(projectId, filter) {
		var params = {
			project: projectId,
			isDeleted: false,
			number: { $gt: 2 }
		};
		_addFilter("priorityId", filter.priorities, params);
		_addFilter("statusId", filter.statuses, params);
		_addFilter("developerId", filter.developers, params);
		_addFilter("testerId", filter.testers, params);
		_addFilter("milestoneId", filter.milestones, params);
		_addFilter("typeId", filter.types, params);
		return params;
	}

	function _addFilter(idProperty, collection, params) {
		if (collection.length > 0)
			params[idProperty] = { $in: collection };
	}

	function _buildSort(direction, comparer) {
		if (comparer == "priority")
			comparer = "priorityOrder";
		else if (comparer == "status")
			comparer = "statusOrder";
		var sort = {};
		sort[comparer] = direction == "ascending" ? 1 : -1;
		sort.number = 1;
		return sort;
	}
};

repository.number = function(projectId, number) {
	return this.one({ project: projectId, number: number });
};

repository.updateIssue = function(model, user) {
	var repositories = require("../repositories");
	return Promise.all([
		repository.details(model._id),
		repositories.Milestone.details(model.milestoneId),
		repositories.Priority.details(model.priorityId),
		repositories.Status.details(model.statusId),
		repositories.IssueType.details(model.typeId),
		repositories.User.details(model.developerId),
		repositories.User.details(model.testerId)
	]).spread(function(issue, milestone, priority, status, type, developer, tester) {
		issue.name = model.name;
		issue.number = model.number;
		issue.details = model.details;
		issue.updatedById = user._id;
		issue.updatedBy = user.name;
		issue.milestoneId = milestone._id;
		issue.milestone = milestone.name;
		issue.milestoneOrder = milestone.order;
		issue.priorityId = priority._id;
		issue.priority = priority.name;
		issue.priorityOrder = priority.order;
		issue.statusId = status._id;
		issue.status = status.name;
		issue.statusOrder = status.order;
		issue.typeId = type._id;
		issue.type = type.name;
		issue.developerId = developer._id;
		issue.developer = developer.name;
		issue.testerId = tester._id;
		issue.tester = tester.name;
		issue.closed = model.closed;
		Promise.promisifyAll(issue).saveAsync();
	});
};

repository.getNextNumber = function(projectId) {
	return repository.one({ project: projectId }, { sort: { number: -1 }}).then(function(issue) {
		return issue.number + 1;
	});
};

repository.issueCountsPerUser = function(projectId) {
	var result = {};
	var model = this.model;
	return require("./userRepository").get({ project: projectId }).then(function(users) {
		return Promise.all(users.map(function(user) {
			return _getCountsForUser(user, model).spread(function(developerCount, testerCount) {
				result[user._id] = {
					developer: developerCount,
					tester: testerCount
				};
			});
		}));
	}).then(function() {
		return result;
	});

	function _getCountsForUser(user, model) {
		return Promise.all([
			model.countAsync({ developerId: user._id }),
			model.countAsync({ testerId: user._id })
		]);
	}
};

module.exports = repository;